### ISSUE TITLE
Cast Youtube?

### ISSUE THREAD

**Issue by Royal2000H** [jaruba/ha-samsungtv-tizen#86]
It's possible to cast to TV (ie, #13 )
```
{
   "entity_id": "media_player.samsungtv",
   "media_content_type": "url",
   "media_content_id": "FILE_URL",
 }
```

And to open the browser:

```
{
  "entity_id": "media_player.samsungtv",
  "media_content_type": "browser",
  "media_content_id": "https://www.google.com",
}
```

Anyone know a way to open the Youtube app to a certain video? The TV must have a way to accept this since it is able to natively cast from my phone's Youtube to the TV. Though granted this may not be possible with the current implementation of this integration.

As a less optimal workaround, perhaps a way to make the integration compatible with [media_extractor](https://www.home-assistant.io/integrations/media_extractor/)?

**Comment by jaruba** [comment]
I don't know of any way to send commands to any TV apps, most likely the YouTube apps (phone and TV) are sending the video through the YouTube servers, not through LAN.

`media_extractor` looks like `youtube-dl` as far as I can tell from it's properties. It should be possible to use `youtube-dl` to extract the the video URL from YouTube web URLs and then send the video URL to the TV to play it. I think Samsung TVs support the video/audio encodings of YouTube, so it should be able to play it.

I can't really offer more help with it though as I never tried to do it myself.

**Comment by obaldius** [comment]
> As a less optimal workaround, perhaps a way to make the integration compatible with [media_extractor](https://www.home-assistant.io/integrations/media_extractor/)?

If manage to do that, please let me know!

**Comment by AKuHAK** [comment]
Btw it is possible to cast youtube links:
https://github.com/balmli/com.samsung.smart/blob/e34b9d5244ec75752f7bb7f89cc83b6a9b87a841/drivers/Samsung/Samsung.js#L226-L242

**Comment by Screwdgeh** [comment]
Thanks for pointing out - this http post command in that code works like a charm! We could implement this as a play_media service call option specifically for youtube videos.

Out of curiosity - what is the use case you are looking to use this for?

**Comment by obaldius** [comment]
Hey Screwdgeh,

could you share the code on how to make that http post command in HA? I'd be very thankful

thanks

**Comment by jaruba** [comment]
Now I'm wondering what other data we could pass to apps... Plex would be my next try..

**Comment by jaruba** [comment]
There seems to be a `/ws/apps/Netflix` too:
- https://github.com/DaveGut/HubitatActive/blob/45be27718e167a9a30e58e68ef94736ac3dbb426/SamsungTvRemote/SamsungTVRemote.groovy#L860-L865

It's mentioned in some other repositories also, but i haven't seen any data being sent to this endpoint yet..

**Comment by jaruba** [comment]
What if the post data is nothing else then the URL path from the official websites? This seems to be true in YouTube's case at least.. i think it might be possible to use the old `/api/v2/` endpoints for other apps too in this way (by posting data to them)

I'm making a lot of presumptions here though..

**Comment by obaldius** [comment]
could somebody please help me?

I'm trying to declare a rest_command in HA but can't get it to work. For now I managed to do:

rest_command:
  video_youtube:
    url: 'http://my_samsungtv_local_ip:8080/ws/apps/YouTube'
    method: POST
    content-type: 'text/plain'

where am I supposed to state the video_id? 

thanks

**Comment by AKuHAK** [comment]
> But we might be stuck with YouTube only support for now, I haven't been able to find other apps being used like this yet and we can't guess the required parameters..

also webbrowser support casting links
`/ws/apps/WebBrowser`

BTW how does anybody know how to launch YouTube with params from commd line using curl?
For now, I can launch youtube via 2 alternate commands:
```
curl -X POST -i http://IP:8001/api/v2/applications/111299001912
curl -X POST http://IP:8001/ws/apps/YouTube
```
and I can close youtube with this command:
```
curl -X DELETE -i http://IP:8001/ws/apps/YouTube/run
```
But I cant figure out how to cast links, I tried:
```
curl -X POST \
  http://IP:8001/ws/apps/YouTube \
  -H 'Content-Type: text/plain' \
  -H 'Content-Length: 13' \
  -d '{"url": "v=3QBKtijv2sk"}'
```
with no luck - maybe I made a mistake in syntax

I have an old 2014 Samsung TV H-series so SmartThings isn't working at all for me.

BTW some apps:
```
8001/ws/apps/emanual
8001/api/v2/applications/20142100003

8001/ws/apps/AdHub
8001/api/v2/applications/111199001564

8001/ws/apps/SSO
8001/api/v2/applications/101400000099

8001/ws/apps/WebBrowser
8001/api/v2/applications/20141000001

8001/ws/apps/CloudPINPage

curl -X POST -i http://IP:8001/api/v2/pincode
curl -X DELETE -i http://IP:8001/api/v2/pincode
```

**Comment by jaruba** [comment]
@AKuHAK 
> But I cant figure out how to cast links

based on the code u initially linked, u should not send `{"url": "v=3QBKtijv2sk"}`, the data that should be posted is only `v=3QBKtijv2sk` (a classic POST, not a POST with JSON)

@obaldius Please be patient, nobody tried it with a rest command in HA's YAML yet

**Comment by Screwdgeh** [comment]
@AKuHAK 

@jaruba hit the nail on the head, this works for me as curl:
```
curl --location --request POST 'http://IPADDRESSOFTV:8080/ws/apps/YouTube' \
--header 'Content-Type: text/plain' \
--data-raw 'v=3QBKtijv2sk'
```

@obaldius - As rest command in HA's configuration.yaml it also works for me (although experience for some reason was not consistent, sometimes it just led me to Youtube app only). It always asks for a confirmation if you want to actually play the video which is quite annoying.

```
rest_command:
  play_youtube_onsamsungtv:
    url: 'http://IPADDRESSOFTV:8080/ws/apps/YouTube'
    method: post
    content_type: 'text/plain'
    payload: 'v=3QBKtijv2sk'
```

**Comment by obaldius** [comment]
Thanks a lot my man, that's working great. Newbie question, how can I execute a curl command in home assistant?

**Comment by Screwdgeh** [comment]
> EDIT: I managed to run the curl with a shell command but I'm getting this confirmation request anyways.

Yes, sorry if I wasn't clear, the confirmation request indeed is always there... seems a YouTube app UX choice - it does that always when trying to launch a video this way... it gets quite hacky but you could program to send the 'enter' key after a sufficient delay if you really want this to work fully automatic...

**Comment by obaldius** [comment]
> hacky but you could program to send the 'enter' key after a sufficient delay if you really want this to work fully automatic...

I just did that, 1 second delay is enough and thanks to jaruba's component the result to me is more than decent, pretty nice, thank you guys

**Comment by pergolafabio** [comment]
Guys, i was testing the script on my older samsung H6400 model (pre-tizen)

```
rest_command:
  video_youtube_fish:
    url: 'http://192.168.0.62:8080/ws/apps/YouTube'
    method: POST
    content_type: 'text/plain'
    payload: 'v=7Y7QaIQET0E'
```
But when i do that, i see the Youtube app starting, so far so good, but then i receive error below... What could be wrong?

![image](https://user-images.githubusercontent.com/43251136/141190142-212c80e1-d65d-48d9-b735-95325e5ae774.png)

**Comment by thematrixdev** [comment]
Is the websocket always there for every TV model? I see an error 404 on port 8080. /ws/apps/YouTube also a 404.

**Comment by ageadss** [comment]
Any news on this? This endpoint ws/apps/YouTube looks completely blocked by the TV, returns 401 and 403
